### ISSUE TITLE
What's the right way to do weight perturbation during model training?

### ISSUE THREAD

**Issue by snsun** [horovod/horovod#1398]
**Environment:**
1. Framework: PyTorch
2. Framework version:
3. Horovod version:
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**

Hi, 

In order to avoid overfitting, I want to perturb the NN's parameters by small noises. But if I applied noises to each rank independently, the parameters of each rank will be different and I am afraid that this will affect the gradient computation. So I want to apply noises to rank 0 only and broad the parameters to other ranks for each rank. What's the right way?
Thanks

**Comment by tgaddair** [comment]
Hey @snsun, do you have a working example using a single process?  The right answer will depend on where / how frequently you're doing the weight perturbation, but in general, the solution would be to run `hvd.broadcast_parameters(model.state_dict(), root_rank=0)` immediately after perturbing the weights.

**Comment by snsun** [comment]
Hi @tgaddair , thanks for your suggestion.  I am trying this method. But I don't know if I run hvd.broadcat_parameters(..., root_rank=0) immediately after perturbing the weights, how can make sure that every rank can obtain the perturbed weights? I think that it is possible that other ranks could run faster than rank0.
